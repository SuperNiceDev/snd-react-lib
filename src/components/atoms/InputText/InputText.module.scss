.root {
  position: relative;
  width: 100%;
  max-width: 100%;
  min-width: 100%;
  margin: 0;
  padding: 0;
  background-color: transparent;
  font-family: inherit;
  font-weight: inherit;
  font-size: inherit;
  line-height: inherit;
  color: inherit;

  border: none;
  border-radius: 0;
  outline: none;
  outline: 0;
  box-shadow: none;

  &.disabled {
    opacity: 0.5;
  }

  &:-webkit-input-placeholder { /* Chrome/Opera/Safari */
    color: inherit;
  }

  &:-moz-placeholder { /* Firefox 19+ */
    color: inherit;
  }

  &:-ms-input-placeholder { /* IE 10+ */
    color: inherit;
  }

  &:-moz-placeholder { /* Firefox 18- */
    color: inherit;
  }

  &:invalid {
    outline: 0;
  }

  &:-webkit-autofill {
    background-color: #FFF !important;
    background-image: none !important;
    box-shadow: 0 0 0px 1000px #FFF inset;
    -webkit-box-shadow: 0 0 0px 1000px #FFF inset;
    -webkit-text-fill-color: #000 !important;
  }

  // .inputTag:-internal-autofill-selected {
  //   color: #000;
  // }
}