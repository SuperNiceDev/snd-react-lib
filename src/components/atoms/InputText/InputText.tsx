import React, {
  InputHTMLAttributes,
  ReactElement,
  useEffect,
  useRef,
} from "react";

import css from "./InputText.module.scss";

export enum InputTypes {
  text = "text",
  password = "password",
  number = "number",
}

// https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete

export enum AutoComplete {
  on = "on",
  off = "off",
  email = "email",
  username = "username",
  name = "name",
  nameGiven = "given-name",
  nameAdditional = "additional-name",
  nameFamily = "family-name",
  streetAddress = "street-address",
  postalCode = "postal-code",
  tel = "tel",
  newPassword = "new-password",
  currentPassword = "current-password",
  currentPasswordConfirm = "current-password-confirm",
}

export type InputTextProps = InputHTMLAttributes<
  HTMLInputElement | HTMLTextAreaElement
> & {
  type?: InputTypes;
  multiline?: boolean;
  focus?: boolean;
  autoComplete?: AutoComplete | string;
  ariaLabel?: string;
};

const InputText = ({
  className,
  type = InputTypes.text,
  name,
  multiline = false,
  focus,
  disabled,
  defaultValue,
  value,
  maxLength,
  placeholder,
  autoComplete,
  ariaLabel,
  onFocus,
  onChange,
  onBlur,
  ...rest
}: InputTextProps): ReactElement => {
  const elemRef = useRef<any>();

  useEffect(() => {
    if (focus) {
      elemRef.current?.focus();
    } else {
      elemRef.current?.blur();
    }
  }, [focus]);

  const Element = !multiline ? "input" : "textarea";
  const clnDisabled = disabled ? ` ${css.disabled} disabled` : "";
  const clnAdd = className ? ` ${className}` : "";
  const cln = `${css.root} InputText${clnDisabled}${clnAdd}`;

  return (
    <Element
      {...rest}
      className={cln}
      ref={elemRef}
      name={name}
      type={type}
      defaultValue={defaultValue}
      value={value}
      maxLength={maxLength}
      disabled={disabled}
      placeholder={placeholder}
      autoComplete={autoComplete}
      aria-label={ariaLabel}
      onFocus={onFocus}
      onChange={onChange}
      onBlur={onBlur}
    />
  );
};

export default InputText;
