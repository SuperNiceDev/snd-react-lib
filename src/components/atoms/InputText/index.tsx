export { default, InputTypes, AutoComplete } from "./InputText";
export type { InputTextProps } from "./InputText";
