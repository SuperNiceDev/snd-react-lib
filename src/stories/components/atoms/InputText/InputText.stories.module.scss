.root {
  padding: 20px;

  .myInput {
    padding: 6px;
    border: 1px solid black;
    
    &.myInputMultiline {
      height: 120px;
    }
  }


  .renderTest {
    padding: 6px;
  }
}
