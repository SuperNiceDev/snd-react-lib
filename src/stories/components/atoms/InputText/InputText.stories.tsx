import React, { useEffect, useState } from "react";

import InputText, {
  InputTextProps,
  InputTypes,
} from "snd-react-lib/components/atoms/InputText";
import { action } from "@storybook/addon-actions";
import type { Meta, StoryObj } from "@storybook/react";

import css from "./InputText.stories.module.scss";

// type InputTextProps = typeof InputText;

export default {
  component: InputText,
  argTypes: {
    type: {
      name: "InputTypes",
      description: "my description",
      control: {
        type: "radio",
      },
      options: [InputTypes.text, InputTypes.password, InputTypes.number],
    },
  },
} as Meta<InputTextProps>;

const Template = (args: InputTextProps) => {
  const { className, multiline } = args;
  const [value, setValue] = useState(args.value);

  useEffect(() => {
    if (args.value !== value) {
      setValue(args.value);
    }
    // eslint-disable-next-line
  }, [args.value]);

  const onChange = (evt: any) => {
    console.log("onChange: ", evt.target.value);
    action("onChange")(evt.target.value);
    setValue(evt.target.value);
  };

  // log("value: ", value);

  const cln = multiline
    ? `${css.myInput} ${css.myInputMultiline} ${className}`
    : `${css.myInput} ${className}`;

  return (
    // <div className={`${css.root}`}>
    <div className={`${css.root} tw:text-red-500 tw:font-bold tw:text-xs`}>
      <span>InputText: </span>
      <InputText {...args} className={cln} value={value} onChange={onChange} />
      <div className={css.renderTest}>Render Test 1</div>
      <div className={css.renderTest}>Render Test 2</div>
      <div className={css.renderTest}>Render Test 3</div>
    </div>
  );
};

type StoryType = StoryObj<InputTextProps>;

const sharedArgs = {
  name: "my_name",
  type: InputTypes.text,
  multiline: false,
  disabled: false,
  focus: true,
  // defaultValue: "my defaultValue",
  value: "my value",
  placeholder: "my placeholder",
  ariaLabel: "my ariaLabel",
  maxLength: 12,
};

export const StoryInputText = {
  name: "InputText",
  render: (args: StoryType) => <Template {...args} />,
  args: {
    ...sharedArgs,
    className:
      "tw:py-1 tw:px-2 tw:text-xs tw:text-lime-500 tw:border tw:border-lime-500",
  },
  parameters: {
    layout: "fullscreen",
  },
};

export const StoryInputTextMultiline = {
  name: "InputText multiline",
  render: (args: InputTextProps) => <Template {...args} />,
  args: {
    ...sharedArgs,
    className:
      "tw:py-1 tw:px-2 tw:text-xs tw:text-lime-500 tw:border tw:border-lime-500",
    multiline: true,
    value: `Lorem Ispum,

Lorem Ispum,

Lorem Ispum,
Lorem Ispum,
Lorem Ispum,

Lorem Ispum
`,
  },
  parameters: {
    layout: "fullscreen",
  },
};
