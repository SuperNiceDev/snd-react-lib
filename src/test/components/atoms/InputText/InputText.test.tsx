import React from "react";

import { fireEvent, render, screen } from "@testing-library/react";
// import '@testing-library/jest-dom/extend-expect';
import userEvent from "@testing-library/user-event";
import InputText from "snd-react-lib/components/atoms/InputText";

const renderComponent = (props = {}) => {
  return render(<InputText {...props} onChange={() => {}} />);
};

describe("InputText Component", () => {
  it("matches the snapshot", () => {
    const { asFragment } = render(<InputText placeholder="My placeholder" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a basic input field by default", () => {
    renderComponent();
    const input = screen.getByRole("textbox");
    expect(input).toBeInTheDocument();
    expect(input).toHaveClass("InputText");
  });

  it("renders a textarea when multiline is true", () => {
    renderComponent({ multiline: true });
    const textarea = screen.getByRole("textbox");
    expect(textarea.tagName).toBe("TEXTAREA");
    expect(textarea).toBeInTheDocument();
  });

  it("applies the provided className", () => {
    renderComponent({ className: "custom-class" });
    const input = screen.getByRole("textbox");
    expect(input).toHaveClass("custom-class");
  });

  it("sets the placeholder attribute", () => {
    const placeholderText = "Enter your name";
    renderComponent({ placeholder: placeholderText });
    const input = screen.getByPlaceholderText(placeholderText);
    expect(input).toBeInTheDocument();
  });

  // it("sets the input type correctly", () => {
  //   renderComponent({ type: InputTypes.password, value: "ffff" });
  //   const input = screen.getByRole("textbox");
  //   expect(input).toHaveAttribute("type", "password");
  // });

  // it("sets the autoComplete attribute", () => {
  //   renderComponent({ autoComplete: AutoComplete.email });
  //   const input = screen.getByRole("textbox");
  //   expect(input).toHaveAttribute("autocomplete", "email");
  // });

  it("sets the aria-label attribute", () => {
    const ariaLabel = "username";
    renderComponent({ ariaLabel });
    const input = screen.getByLabelText(ariaLabel);
    expect(input).toBeInTheDocument();
  });

  it("sets the input as disabled", () => {
    renderComponent({ disabled: true });
    const input = screen.getByRole("textbox");
    expect(input).toBeDisabled();
    expect(input).toHaveClass("disabled");
  });

  it("focuses the input when focus prop is true", () => {
    renderComponent({ focus: true });
    const input = screen.getByRole("textbox");
    expect(input).toHaveFocus();
  });

  it("triggers onFocus event handler", () => {
    const onFocusMock = jest.fn();
    renderComponent({ onFocus: onFocusMock });
    const input = screen.getByRole("textbox");
    fireEvent.focus(input);
    expect(onFocusMock).toHaveBeenCalledTimes(1);
  });

  it("triggers onChange event handler", async () => {
    // const onChangeMock = jest.fn();
    // renderComponent({ onChange: onChangeMock });
    // const input = screen.getByRole("textbox");
    // fireEvent.change(input, { target: { value: "test" } });
    // expect(onChangeMock).toHaveBeenCalledTimes(1);
    //
    const onChange = jest.fn();
    renderComponent({ onChange });
    const input = screen.getByRole("textbox");
    input.focus();
    const user = userEvent.setup();
    await user.type(input, "Hello, World!");
    expect(input).toHaveValue("Hello, World!");
  });

  it("triggers onBlur event handler", () => {
    const onBlurMock = jest.fn();
    renderComponent({ onBlur: onBlurMock });
    const input = screen.getByRole("textbox");
    fireEvent.blur(input);
    expect(onBlurMock).toHaveBeenCalledTimes(1);
  });

  it("sets the defaultValue correctly", () => {
    renderComponent({ defaultValue: "default value" });
    const input = screen.getByRole("textbox");
    expect(input).toHaveValue("default value");
  });

  it("sets the value correctly", () => {
    renderComponent({ value: "controlled value" });
    const input = screen.getByRole("textbox");
    expect(input).toHaveValue("controlled value");
  });

  it("sets the maxLength attribute", () => {
    renderComponent({ maxLength: 10 });
    const input = screen.getByRole("textbox");
    expect(input).toHaveAttribute("maxLength", "10");
  });
});
