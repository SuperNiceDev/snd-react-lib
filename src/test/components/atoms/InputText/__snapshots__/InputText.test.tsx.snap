// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputText Component matches the snapshot 1`] = `
<DocumentFragment>
  <input
    class="root InputText"
    placeholder="My placeholder"
    type="text"
  />
</DocumentFragment>
`;
